Else check

Oakheart27 opened the pull request:
Else Branch

Oakheart27 committed:
Add files via upload

Just to show you what I had in mind. The first solution was just to try and have a less complicated solution.

Oakheart27 committed:
Update blah.cpp

Oakheart27 committed:
Update blah.cpp

evdru reviewed (commented):
Using standard libraries such as stdbool.h, cstring.h, cctype.h.

Maybe put function declarations in a seperate header file.

Have a separate driver file where main is located to bootstrap the program.

Recommend to have a make file to compile and run your program. As well as include a README with instructions on how to compile and run your program.

Include simple unit tests to make sure you handle the base case and edge cases.

evdru commented:
Maybe put function declarations in a seperate header file.

evdru commented:
Have a separate driver file where main is located to bootstrap the program.

evdru commented:
Recommend to have a make file to compile and run your program. As well as include a README with instructions on how to compile and run your program.

evdru commented:
Include simple unit tests to make sure you handle the base case and edge cases.

evdru commented on blah.cpp:
```diff
@@ -0,0 +1,82 @@
+    numerator = atoi(mantissa_value_string);
+    //printf("The mantissa numerator is: %i\n", numerator);
+
+    return numerator;
```
bool mantissa() returns type int* and not bool?

clockian-NAU replied:
Just to clarify this a bit further, it seems like C++ is similar to C in that bools don't actually exist. It's just an int that holds either 0 or 1. It also seems like the bool can, maybe implicitly, be converted to a number.

Given the above, it seems like in the line
```numerator = atoi(mantissa_value_string);```
This assigns the output to the reference parameter numerator, which acts kind of like a pointer in C, but with more restrictions on it, as C++ is designed(as far as I know) to de-emphasize pointer manipulation that is rampant in C. For this function, this is how the values are returned out of the function, and how the denominator is also suppose to be returned.

Now moving to 
```return numerator;```
What's happening here is that you're trying to return the reference parameter out of a function with the bool return type. This should error, but what seems to be happening is that the return dereferences the numerator and looks at the value it holds, which is an int. It takes the int and tries to fit it in a bool, which is just an int that can only be 0 or 1, thus the number gets reduced down to 1 which is then returned.

I recommend refactoring this code so that the return is a separate variable that solely determines if the function error'd or not, and just passing out the numerator and denominator values through the reference parameters. This also applies to the characteristic function.

clockian-NAU commented:
Just to expand off of what Evan has said.

* "Using standard libraries such as stdbool.h, cstring.h, cctype.h."
Considering this is a very limited embedded system this program is to run on, this is a very reasonable assumption to make. You may need to inquire as to the specific capabilities of the system before using even standard libraries. That being said, this is also a good idea to start development, so you have more tools to work with, and if you need to remove a tool, you know what you need to replace it with. Thanks for that idea.

* "Maybe put function declarations in a separate header file."
For me personally, I use header files as a quick reference, rather than looking at the code itself, so I find them very useful.

* "Have a separate driver file where main is located to bootstrap the program."
This is just standard good practice, but it would also allow you to utilize the OO nature of C++ more easily, a very powerful tool.

* "Recommended to have a make file to compile and run your program. As well as include a README with instructions on how to compile and run your program."
Compilation instructions are nice(personally I had the problem of forgetting the .cpp >.<) but makefiles are great! If you want to get fancy, you can have your project build with just make, and have a separate testing suite build with make test. It can really speed development within your team if you make it super easy for everyone to compile on a whim.

* "Include simple unit tests to make sure you handle the base case and edge cases."
This is a personal thing with me and C, but I find I can't write good, or at least usable, code without tests. Granted, your mileage may vary, and given the limited scope of this project, it may not be necessary, but it is something to consider. It doesn't even have to be anything fancy, it could just be a bunch of functions to print the output of your program with different inputs.

clockian-NAU commented:
As for the code...

A file header would be nice, as sometimes I work on Vim, a cmd line editor, due to the limited capabilities of my laptop, and it doesn't show the file name I'm working on very easily. It doesn't need to be fancy, but the following would be appreciated.
```c++
/*
 * @file <file-name>
 *
 * @author [maybe, not really necessary for this project I think]
 *
 * @brief <small description of the purpose of the file>
 */
```

clockian-NAU commented:
With regards to comments, if would be great if you could have them follow a standard format for the code. Below is what I personally prefer, but as long as all of them follow a similar format, anything works honestly.
```C++
// Capitalized first letter with a space from the double slashes

someExampleCode(cool);  // Two spaces away from the code for inline

clockian-NAU commented:
With regards to the variable naming, good job on having descriptive variables names. It would be great if you could change your variable names to all be either snake case or camel case though, this is some consistency issues with that.

clockian-NAU commented:
With regards to the program's output, you are missing a requirement where you need to output the denominator in the mantissa function, as it only outputs the numerator at the moment.

clockian-NAU commented:
Last of the general project and code remarks, I took this opportunity to run the tests I constructed for this assignment and more on your code, below are the results

## Tests that are suppose to succeed

#### Successful tests, tests that passed
```C
"123.456"
"-123.456"
"    -2.3"
"-0002.3"
"    -0002.3"
"2.3      "
```
#### Error'd tests, that are suppose to pass
```C
"2.0"
"2."
"2" - This causes stack smashing
"-2"
"0.3"
"    0.3"
".3"
"2."
"-.3"
```
#### Tests that output wrong numbers
```
"2.3000" -> outputs: c = 2, n = 3000 | n should be 3
"2.3000    " -> see above ^
```

## Tests that are suppose to Fail

#### Successful test, tests that error'd correctly
```
"a.1"
"-a.1"
"-0a.1"
"1.a1"
"1.a0"
"     "
"" -> This exits the program with no output, 
        I don't know whether this is intended behavior or not
```

#### Failed tests, tests that output when they are suppose to fail
```
"-1a.1" -> c = -1, n = 1
"1.1a" -> c = 1, n = 1
"1.1.1" -> c = 1, n = 1
"1.+1" -> c = 1, n = 1
"1.-1" -> c = 1, n = -1
```

clockian-NAU commented on blah.cpp:
```diff
@@ -0,0 +1,82 @@
+        // Checks if there is another symbol besides a digit or a + - symbol
+        for (int i = 0; i < strlen(number); i = i + 1)
+        { 
+          if(number[i] >= '0' || number[i] <= '9' && number[i] != '+' && number[i] != '-')
```
You may want to consider splitting the multiple conditions apart so that it is easier to read at a glance. You may want to reconsider the purpose of this while loop and consider if having a separate loop for '-' and '+' maybe better. You may want to consider if testing for '+' is a part of the requirements of the project

clockian-NAU commented on blah.cpp:
```diff
@@ -0,0 +1,82 @@
+bool characteristic(char numString[], int& c)
+{
+    //turn char into int
+    c = atoi(numString);
```
While I do like the simplicity of this function, and while I am surprised at the mileage atoi() gives to this function, it is a bit to simple for this assignment, failing to cover some edge cases.

clockian-NAU commented on blah.cpp:
```diff
@@ -0,0 +1,82 @@
+    while (numString[currentIndex] != '\0')
+    {
+      mantissa_value_string[currentIndex - offset] = numString[currentIndex];
+      currentIndex++;
```
The indentation in this loop is off by 2 spaces

clockian-NAU commented on blah.cpp:
```diff
@@ -0,0 +1,82 @@
+
+    while (numString[currentIndex] != '\0')
+    {
+      mantissa_value_string[currentIndex - offset] = numString[currentIndex];
```
Consider removing the obfuscating math, as the currentIndex minus the offset just counts from 0 to the life of the loop. Might as well us an i++

clockian-NAU commented on blah.cpp:
```diff
@@ -0,0 +1,82 @@
+        //do some math with c, n, and d
+        printf("test true \n");
+        printf("The characteristic is: %i\n", c);
+        printf("The mantissa numerator is: %i\n", n);
```
I originally thought printf() shouldn't work in C++, since it's C, but it turns out I'm wrong. I would personally advocate for using std::cout just because this is C++, and you aren't outputting enough IO for the performance issue with cout to matter. That being said, here the interesting StackOverflow thread that showed me this. https://stackoverflow.com/questions/2872543/printf-vs-cout-in-c

KennySprings4Life commented on blah.cpp:
```diff
@@ -0,0 +1,82 @@
+    int c, n, d;
+
+    //if the conversion from C string to integers can take place
+    if(characteristic(number, c) && mantissa(number, n, d))
```
You should consider writing separate functions instead of having all of the functionality occur directly within the main function. The main should act more as a driver which calls other functions, instead of handling all of the processing by itself.
For example: 
Main takes the input from the user, but then sends it to another function that "sanitizes" the data by removing trailing white spaces, verifies that the data is indeed valid, and perhaps strips the "-" / "+" sign from the input for later use. The main would then send this "clean" data to the other functions which will then perform the actions that your current mantissa and characteristic functions do.

KennySprings4Life commented:
> 
> 
> With regards to comments, if would be great if you could have them follow a standard format for the code. Below is what I personally prefer, but as long as all of them follow a similar format, anything works honestly.
> 
> ```c++
> // Capitalized first letter with a space from the double slashes
> 
> someExampleCode(cool);  // Two spaces away from the code for inline
> ```

I would agree with Jasque here. I noticed also some of the comments for this pull request were capitalized, while others were not. Don't forget to proofread your comments, as well as your code.

Oakheart27 committed:
Worked on Suggested Code Edits

Standardized variable names. Broke up else check branch. Moved func1.cpp into several different files including main, make file, func1cpp, and header file. Fixed indentation. Added comments.

morganlovato reviewed (approved):
looks good

jdm568 committed:
Fixed some errors

Just some debugging so I can test it.

Oakheart27 committed:
String to int conversion without atoi

Code to swich string to int. Logic is commented. It's here if you guys think we should use it. I don't know if you guys want to stick withall of the code seperate or go back to one file where everything worked. Either way the code can just be copied and pasted.

jdm568 reviewed (approved):
Looks good!

morganlovato reviewed (approved):
Looks good to me too!
